common/test/common_tests/logic/swap_and_reset_test.cljc

```diff
@@ -0,0 +1,175 @@
+;;
+;; Copyright (c) KALEIDOS INC
+
+(ns common-tests.logic.swap-and-reset-test
+  (:require
+   [app.common.files.changes-builder :as pcb]
+   [app.common.files.libraries-helpers :as cflh]
+   [app.common.pprint :as pp]
+   [app.common.types.component :as ctk]
+   [app.common.types.file :as  ctf]
+   [app.common.uuid :as uuid]
+   [clojure.test :as t]
+   [common-tests.helpers.compositions :as thc]
+   [common-tests.helpers.files :as thf]
+   [common-tests.helpers.ids-map :as thi]))
+
+(t/use-fixtures :each thi/test-fixture)
+
+;; Related .penpot file: common/test/cases/swap-and-reset.penpot
+(t/deftest test-simple-swap
+  (let [;; Setup
+        file
+        (-> (thf/sample-file :file1)
+            (thc/add-simple-component-with-copy :component-1 :component-1-main-root :component-1-main-child :component-1-copy-root)
+            (thc/add-simple-component :component-2 :component-2-root :component-2-child))
+
+        component-1-copy-root (thf/get-shape file :component-1-copy-root)
+        component-1           (thf/get-component file :component-1)
+        component-2           (thf/get-component file :component-2)
+        page                  (thf/current-page file)
+
+        ;; Action
+        [new-shape all-parents changes]
+        (cflh/generate-component-swap (pcb/empty-changes)
+                                      (:objects page)
+                                      component-1-copy-root
+                                      (:data file)
+                                      page
+                                      {(:id  file) file}
+                                      (:id component-2)
+                                      0
+                                      nil
+                                      {})
+
+        file' (thf/apply-changes file changes)
+
+        ;; Get 
+        swapped (thf/get-shape-by-id file' (:id new-shape))]
+
+    ;; Check
+    (t/is (not= (:component-id component-1-copy-root) (:component-id swapped)))
+    (t/is (= (:id component-2) (:component-id swapped)))
+    (t/is (= (:id file) (:component-file swapped)))))
+
+(t/deftest test-swap-anidated
+  (let [;; Setup
+        file
+        (-> (thf/sample-file :file1)
+            (thc/add-simple-component :component-1 :component-1-main-root :component-1-main-child)
+            (thc/add-frame :component-container)
+            (thf/instantiate-component :component-1 :component-1-copy-root {:parent-label :component-container})
+            (thf/make-component :component-container-main :component-container)
+            (thf/instantiate-component :component-container-main :component-container-instance)
+            (thc/add-simple-component :component-2 :component-2-main-root :component-2-main-child))
+
+        page        (thf/current-page file)
+        component-1 (thf/get-component file :component-1)
+        component-2 (thf/get-component file :component-2)
+        component-3 (thf/get-component file :component-3)
+
+        copy
```

Comment:
I've just add a `:children-labels` parameter on instantiate-component to make this easier